src/abi/substrate.rs

```diff
@@ -1,437 +1,473 @@
+    let scalety = match ty {
+        // Substrate doesn't like primitive types which aren't a power of 2
+        // The abi encoder/decoder fixes this automatically
+        ast::Type::Uint(n) => format!("u{}", n.next_power_of_two()),
+        ast::Type::Int(n) => format!("i{}", n.next_power_of_two()),
+        _ => unreachable!(),
+    };
 
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-pub struct Array {
-    len: usize,
-    #[serde(rename = "type")]
-    ty: usize,
-}
+    let def = match scalety.as_str() {
+        "u8" => TypeDefPrimitive::U8,
+        "u16" => TypeDefPrimitive::U16,
+        "u32" => TypeDefPrimitive::U32,
+        "u64" => TypeDefPrimitive::U64,
+        "u128" => TypeDefPrimitive::U128,
+        "u256" => TypeDefPrimitive::U256,
+        "i8" => TypeDefPrimitive::I8,
+        "i16" => TypeDefPrimitive::I16,
+        "i32" => TypeDefPrimitive::I32,
+        "i64" => TypeDefPrimitive::I64,
+        "i128" => TypeDefPrimitive::I128,
+        "i256" => TypeDefPrimitive::I256,
+        _ => unreachable!(),
+    };
 
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-pub struct SequenceDef {
-    sequence: Sequence,
-}
+    let ty = Type::new(
+        Default::default(),
+        vec![],
+        TypeDef::Primitive(def),
+        Default::default(),
+    );
 
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-pub struct Sequence {
-    #[serde(rename = "type")]
-    ty: usize,
-}
+    get_or_register_ty(&ty, registry);
 
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-#[serde(untagged)]
-enum Type {
-    Builtin { def: PrimitiveDef },
-    BuiltinArray { def: ArrayDef },
-    BuiltinSequence { def: SequenceDef },
-    Struct { path: Vec<String>, def: Composite },
-    Enum { path: Vec<String>, def: EnumDef },
+    ty
 }
 
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-struct BuiltinType {
-    id: String,
-    def: String,
-}
+type Cache = HashMap<ast::Type, PortableType>;
 
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-struct EnumVariant {
-    name: String,
-    discriminant: usize,
-}
+/// given an `ast::Type`, find and register the `scale_info::Type` definition in the `PortableRegistry`
+fn resolve_ast(
+    ty: &ast::Type,
+    ns: &ast::Namespace,
+    registry: &mut PortableRegistry,
+    cache: &mut Cache,
+) -> PortableType {
+    // early return if already cached
+    if let Some(ty) = cache.get(ty) {
+        return ty.clone();
+    }
 
-#[derive(Deserialize, Serialize, PartialEq, Eq)]
-struct EnumDef {
-    variant: Enum,
-}
+    match ty {
+        //  should reflect address_length for different substrate runtime
+        ast::Type::Address(_) | ast::Type::Contract(_) => {
+            // substituted to [u8 ;address_length]
+            let address_ty = resolve_ast(
+                &ast::Type::Array(
+                    Box::new(ast::Type::Uint(8)),
+                    vec![ArrayLength::Fixed(
+                        BigInt::from_u8(ns.address_length as u8).unwrap(),
```

Comment:
I *think* this can be written without a cast: `BigInt::from(ns.address_length).unwrap()`